backend/tests/test_sso.py

```diff
@@ -0,0 +1,206 @@
+            'SAMLResponse': self.get_sso_response(),
+        }
+        with self.client.session_transaction() as sess:
+            self.assertFalse(sess.get('MO-Token'))
+
+        r = self.client.post('/saml/acs/', data=data)
+
+        samlUserData = {
+            'urn:oid:0.9.2342.19200300.100.1.1': ['bruce'],
+            'urn:oid:0.9.2342.19200300.100.1.3': ['bruce@kung.fu'],
+            'urn:oid:2.5.4.41': ['Bruce Lee']
+        }
+        samlNameId = '_e3dfaf3e3385fd182b1c4d4164644393cce3ac7bfe'
+        samlSessionIndex = '_6a54ed11e21b64af1a0380b1fba3ec575b05855465'
+
+        with self.client.session_transaction() as sess:
+            self.assertTrue(sess.get('MO-Token'))
+            self.assertEqual(samlUserData, sess.get('samlUserdata'))
+            self.assertEqual(samlNameId, sess.get('samlNameId'))
+            self.assertEqual(samlSessionIndex, sess.get('samlSessionIndex'))
+
+    @patch('onelogin.saml2.auth.OneLogin_Saml2_Auth.get_errors',
+           lambda *x, **y: ['ERROR'])
+    def test_acs_returns_errors(self):
+        data = {
+            'SAMLResponse': self.get_sso_response(),
+        }
+
+        expected = {
+            'description': "['ERROR']",
+            'error': True,
+            'error_key': 'E_UNKNOWN',
+            'status': 500
+        }
+
+        r = self.client.post('/saml/acs/', data=data)
+
+        self.assertEqual(expected, r.json)
+
+    def test_slo_redirects_to_logout_page(self):
+        r = self.client.get('/saml/slo/')
+        url = urlparse(r.location)
+
+        self.assertEqual(302, r.status_code)
+        self.assertEqual('https', url.scheme)
+        self.assertEqual('192.168.1.212', url.netloc)
+        self.assertEqual('/simplesaml/saml2/idp/SingleLogoutService.php',
+                         url.path)
+
+    def test_sls_redirects_correctly(self):
+        data = {
+            'SAMLResponse': self.get_slo_response(),
+        }
+        r = self.client.get('/saml/sls/', query_string=urlencode(data))
+
+        self.assertEqual(302, r.status_code)
+        self.assertEqual('http://127.0.0.1:5000/', r.location)
+
+    def test_sls_deletes_session(self):
+        data = {
+            'SAMLResponse': self.get_slo_response(),
+        }
+
+        with self.client.session_transaction() as sess:
+            sess['MO-Token'] = "TOKEN"
+
+        r = self.client.get('/saml/sls/', query_string=urlencode(data))
+
+        with self.client.session_transaction() as sess:
+            self.assertFalse(sess)
+
+    @patch('onelogin.saml2.auth.OneLogin_Saml2_Auth.get_errors',
+           lambda *x, **y: ['ERROR 2'])
+    def test_sls_returns_errors(self):
+        data = {
+            'SAMLResponse': self.get_slo_response(),
+        }
+
+        r = self.client.get('/saml/sls/', query_string=urlencode(data))
+
+        expected = {
+            'description': "['ERROR 2']",
+            'error': True,
+            'error_key': 'E_UNKNOWN',
+            'status': 500
```

Comment:
Unfortunately, we cannot assume that a SAML error is, in fact, an error condition. For example, the cause of the error might simply be an incorrect login — see `tests/mocking/auth/adfs-failed-login.xml`, for example.

I think we should return 401 or 403 — or whatever we did previously — when the login fails.

Reply:
SSO works differently than what we used before, so if a user fails to authenticate they'll be stopped by the IdP, and thus we'll never receieve a failed login response such as `tests/mocking/auth/adfs-failed-login.xml`

Reply:
Ah, that's good news — can you think of anything else that might trigger this? Did you see it during manual testing?